### ISSUE TITLE
Support Sha256 for release hash

### ISSUE THREAD

**Issue by Daniel-Svensson** [velopack/velopack#105]
Would you be open to add support for SHA256 hashes in releases.*.json ?

Use case:
We want to make velopack updates "tamper prof" and is looking to use digital signature to protect updates.
We estimate that it is doable by either using nuget signatures (or maybe detached CMS/PKCS#7 signature since .nupkg might go away) or by signing the release files, but for this case to be deemed fully "secure" and tamper proof we need to use SHA256 for .nupkg verification.

**Note**: We are currently evaluating velopack as a complement to our clickonce based solution and if we decide to go ahead we are open to help with the implementation (The C# part at least) and/or sponsoring the project. 

Proposed solution:
* vpk pack adds both "SHA1" and "SHA256" hashes (so it is compatible with older clients)
   * An alternative would be to add a parameter to specify hashing algorithm(s), but that sounds more complex to implement 
* velopack client reads and checks files using SHA256 (if availible)
   *  VelopackAsset class is extended with a new field  `public string SHA256 { get; init; }`
   * If SHA256 is not null then it is used instead of SHA1

**Comment by caesay** [comment]
I'm not opposed to SHA256 hashes, but the hash in the releases.json is not a security measure, it's just to verify that the file was downloaded correctly. For example, this doesn't protect you from a MITM attack where someone can provide their own releases.json / .nupkg files - or in the scenario that your upload server is compromised and an attacker can upload malware

**Comment by Daniel-Svensson** [comment]
Thanks for the swift reply, I will probably know how we will do in a week or two.

> For example, this doesn't protect you from a MITM attack where someone can provide their own releases.json / .nupkg files - or in the scenario that your upload server is compromised and an attacker can upload malware

That is why we would sign the release feed (and serve it from a more trusted location).
We would probably write a custom IUpdateSource that checks digital signatures on download so we might maybe end up using the same logic and add signatures to the packages as well.

**Comment by caesay** [comment]
I would like package signing/verification added as a feature but it is complex and there are a lot of considerations. There were some (potentially outdated) ideas here https://github.com/clowd/Clowd.Squirrel/issues/110. If you have some concrete ideas for viable package signing I would be happy to discuss it and potentially we can make it officially supported. Feel free to join our Discord and ping me if you'd like to have a chat about this.

**Comment by caesay** [comment]
Hi, any update on your thoughts here and/or updates on your move to Velopack?

**Comment by Daniel-Svensson** [comment]
I have implemented the updater, except  signatures at the moment and we will try to collect some feedback from some of our end user before making the last polishing.

So far we are happy (except that fsecure/withsecure deepguard make full updates take 30-50s instead of seconds) and I we will add signing and Sha256 hashes before any broader rollout.


The current plan is to write a custom update source that verifies signatures on releases (using cms signature with code signing, or maybe something similar to jwt with a  static hardcoded key).

We will realistically look at sha256 sometime around September (maybe faster)

**Comment by Daniel-Svensson** [comment]
Ps: I will try to remember to submit a sample project with some kind of signature check when done

**Comment by caesay** [comment]
> except that fsecure/withsecure deepguard make full updates take 30-50s instead of seconds
Is there changes we could make to Velopack improve the situation?

**Comment by Deadpikle** [comment]
Chiming in here (maintainer of NetSparkle; Velopack is a great repo/software bundle you have going here and I'm hoping to learn some things from you!):

We use [BouncyCastle](https://github.com/bcgit/bc-csharp) for ed25519 signatures, which is the same sort of signature check that [Sparkle](https://github.com/sparkle-project/Sparkle) uses. Ed25519 is a public/private key scheme, which should help alleviate some of the problems you've mentioned above. Unfortunately the size of the newer versions of BC is rather large (see https://github.com/bcgit/bc-csharp/issues/451 and https://github.com/bcgit/bc-csharp/issues/422; although it looks like recently trimming has been added which should help the file size greatly. The size issues are a TBD on our side and need to be looked at; perhaps a different ed25519 library can be used or we just tell our users to use trimming or we trim on our side or something else.)

You can see the verification code on the client side here: https://github.com/NetSparkleUpdater/NetSparkle/blob/develop/src/NetSparkle/SignatureVerifiers/Ed25519Checker.cs — includes code for streaming of files to save on RAM.

And the code for creating the public/private keys as well as more verification as part of our CLI tool: https://github.com/NetSparkleUpdater/NetSparkle/blob/develop/src/NetSparkle.Tools.AppCastGenerator/SignatureManager.cs — and in looking at that I realize we aren't doing file streaming there to verify the signature; that should be added in. But you should see what we're doing between those two files.

Basically, on the generator side, the user uses the app cast CLI tool to create the release feed based on their already-existing software installer. (Our package doesn't handle installers.) The CLI tool generates the release feed, which contains signatures, as well as a signature for the release feed itself. Then the client, which ships with the user's public key, can look at these signatures on downloading the release feed/etc. to make sure that everything is legitimate. (We offer options to skip signature checking, but, well, you know, that's done at the user's risk.)

If you have any questions, let me know!
